Java File API

vinopm opened the pull request:
File API functionality implemented in Java & Travis CI enabled.

vinopm commented:
@johnnonweiler

Commit:
Initial commit for Java File API

Commit:
Add gradle wrapper

rtblackwell commented on data_pipeline_api/java_api/src/main/java/uk/ramp/api/FileApi.java:
```diff
@@ -0,0 +1,98 @@
+   * @param query input query
+   */
+  public FileChannel openForRead(MetadataItem query) throws IOException {
+    var overridenQuery = overridesApplier.applyReadOverrides(query);
```
trivial: overridden

vinopm replied:
Will update, thanks

rtblackwell commented on data_pipeline_api/java_api/src/main/java/uk/ramp/config/ConfigReader.java:
```diff
@@ -0,0 +1,32 @@
+import uk.ramp.yaml.YamlReader;
+
+class ConfigReader implements Reader<ImmutableConfig> {
+  private static final String LOCATION = "config.yaml";
```
not sure if we'd expect this to be constant or an arg to the model

vinopm replied:
Right now the FileApi class takes the directory (which contains config.yaml) as an input parameter. Happy to move this up to the API level and have the config.yaml path passed in instead of the directory. Let me know. Either way, it should be trivial to change this at a later stage or in a follow up PR if we're not sure yet

vinopm replied:
I've changed this now so that the actual config file path is passed in as an arg to the File API

rtblackwell commented on data_pipeline_api/java_api/src/main/java/uk/ramp/access/AccessLog.java:
```diff
@@ -0,0 +1,31 @@
+package uk.ramp.access;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
```
probably a stupid question, but I'm curious, what's the relationship between YAML and JSON in this API, my speculation is that the yaml segments are being split into json like things that the json deserializer can then make into objects?

vinopm replied:
This is a perfectly sensible question to ask. These Jackson core annotation names are a bit misleading. Even though these annotations have 'Json' their name, it will work. The Jackson ObjectMapper can be configured to parse all sorts of different formats including Yaml. In our case we are using the YAMLMapper (which is a configuration of ObjectMapper for Yaml). Given all of this, for yaml mappings you would still use the core annotations which have 'Json' in their names.

rtblackwell commented on data_pipeline_api/java_api/src/main/java/uk/ramp/access/AccessLoggerImpl.java:
```diff
@@ -0,0 +1,66 @@
+    var accessEntry =
+        ImmutableAccessEntry.builder()
+            .type("write")
+            .timestamp(clock.instant())
```
How does this timestamp get logged out to access.yaml, as I guess instant is millis since epoch? The upload/download scripts are currently in python, so we need to make sure there's compatibility

vinopm replied:
Good question, and definitely one we should discuss. 

```
timestamp:
  nano: 134173000
  epochSecond: 1594376918
```

This is how it is currently logged out. Happy to change this to a standard format. Let me know

rtblackwell replied:
https://stackoverflow.com/a/45674593

I think if you do this and have it output the "2017-08-14T12:17:47.720Z" format then that's interpretable by the python yaml loader and I assume will also be interpretable by your code without more work than that configuration?

vinopm replied:
That's a nice approach. Will update, thanks.

vinopm replied:
@rtblackwell Made the suggested change and by default it outputs in this format: "2020-07-10T14:08:10.367415Z". Would this be ok, or is it necessary for the seconds/millis be truncated to 3 digits?

rtblackwell replied:
I think that will work fine
```
import yaml
yaml.safe_load("2020-07-10T14:08:10.367415Z")
datetime.datetime(2020, 7, 10, 14, 8, 10, 367415, tzinfo=datetime.timezone.utc)
```

Commit:
Publish to github packages if branch is master

Commit:
Fix gradle publish

Commit:
Fix gradle publish

Commit:
Fix gradle publish - keep it simple for now.

mrow84 commented on data_pipeline_api/java_api/src/main/java/uk/ramp/api/FileApi.java:
```diff
@@ -0,0 +1,98 @@
+    var normalisedFilename = fileDirectoryNormaliser.normalisePath(filename);
+    Files.createDirectories(Path.of(fileDirectoryNormaliser.parentPath()));
+    Files.createFile(Path.of(normalisedFilename));
+    accessLogger.logWrite(query, overridenQuery);
```
I've only just scanned this, and I'm not very familiar with modern Java so maybe I've missed something, but does the write logging happen _before_ the write, or once the file has been closed? It should be the latter, and the hash of the resulting file should be computed. Again, apologies if I've just missed something.

vinopm replied:
A very good question and well spotted. You're correct. The 'write logging' happens before the write to file actually happens. We log that there has been a write access just before returning the file handle. 

I didn't realise that this was a strict requirement. This will require some fiddling around as Java doesn't let you directly override the close() method on the file handle (FileChannel objects). 

The point you mentioned about the hash of the resulting file needing to be computed again is important. 

I'll work on getting this functionality implemented. 

Thanks

mrow84 replied:
Yeah nw, it's probably the most subtle/awkward part of how this works, which is why I went looking for it. I still have to look through the rest, but I imagine it's basically fine.

Commit:
Fix typo

johnnonweiler commented on data_pipeline_api/java_api/src/main/java/uk/ramp/access/AccessLoggerFactory.java:
```diff
@@ -0,0 +1,32 @@
+    try {
+      underlyingWriter = Files.newBufferedWriter(Path.of(accessLogPath));
+    } catch (IOException e) {
+      throw new UncheckedIOException(new IOException(e));
```
Why do you create a new IOException here?  (I have a feeling that perhaps I once knew the answer to this question, but can't remember!)

vinopm replied:
Woops, good spot, definitely a mistake! I'll update. Thanks!

johnnonweiler commented on data_pipeline_api/java_api/src/test/java/uk/ramp/metadata/HashVerificationMetadataSelectorTest.java:
```diff
@@ -0,0 +1,33 @@
+    when(metadataSelector.find(query)).thenReturn(query);
+    when(hasher.hash("file1")).thenReturn("hash1");
+
+    var hashVerificationFileApi =
```
Curious choice of variable name.

vinopm replied:
I must have missed out renaming this when refactoring my own code :). Thanks for spotting, I will update it.

johnnonweiler commented on data_pipeline_api/java_api/src/test/java/uk/ramp/metadata/HashVerificationMetadataSelectorTest.java:
```diff
@@ -0,0 +1,33 @@
+        new HashVerificationMetadataSelector(metadataSelector, hasher, true);
+
+    assertThat(hashVerificationFileApi.find(query)).isEqualTo(query.withCalculatedHash("hash1"));
+  }
```
Is there a test somewhere with an invalid hash?

johnnonweiler replied:
Also, you're not actually testing anywhere that you're using SHA-1 hashes, which could be important, but I guess it would quickly become obvious if you got that wrong, as soon as anyone starts using the API.

vinopm replied:
Thanks John, I'll add tests in for both your points mentioned.

johnnonweiler commented on data_pipeline_api/java_api/src/main/java/uk/ramp/api/FileApi.java:
```diff
@@ -0,0 +1,98 @@
+    var overriddenQuery = overridesApplier.applyWriteOverrides(query);
+    var filename = overriddenQuery.filename().orElseThrow();
+    var normalisedFilename = fileDirectoryNormaliser.normalisePath(filename);
+    Files.createDirectories(Path.of(fileDirectoryNormaliser.parentPath()));
+    Files.createFile(Path.of(normalisedFilename));
```
Could `normalisedFilename` be a path to a file in another folder which does not exist, or is there a guarantee that all files will be written to the same folder (given by `fileDirectoryNormaliser.parentPath()`)?

vinopm replied:
I will change this to:
```
Files.createDirectories(Path.of(normalisedFilename).getParent());
Files.createFile(Path.of(normalisedFilename));
```
to ensure that we're creating the directories based on the parent folder of the normalisedFilename (which could point to a different path than what is stored as the parent path in fileDirectoryNormaliser)

johnnonweiler commented on data_pipeline_api/java_api/src/main/java/uk/ramp/config/ConfigFactory.java:
```diff
@@ -0,0 +1,30 @@
+      FileReader fileReader,
+      Instant openTimestamp,
+      FileDirectoryNormaliser fileDirectoryNormaliser) {
+    var config = new ConfigReader(yamlReader, fileDirectoryNormaliser).read();
```
It would make more sense to me to pass the normalised path to the ConfigReader constructor (instead of `fileDirectoryNormaliser`).  This `LOCATION` is hard coded in both ConfigFactory and ConfigReader, which seems wrong.

vinopm replied:
Good spot, I agree. I will update this. Thanks John

johnnonweiler commented on data_pipeline_api/java_api/src/main/java/uk/ramp/hash/HasherFactory.java:
```diff
@@ -0,0 +1,18 @@
+
+  public Hasher fileHasher(FileReader fileReader, Instant openTimestamp) {
+    return fileName -> new Sha1Hasher().hash(fileReader.read(fileName) + openTimestamp.toString());
+  }
```
I find these fileHasher methods a bit tricky to follow.  You're returning a lambda expression, the method's return type is Hasher, and the lambda expression is using a Hasher.  Since (as far as I can tell) only one of these fileHasher methods is ever used, and only once, I suggest removing them.

vinopm replied:
Agreed, I've simplified and updated. Please let me know if you think it's still a bit tricky to follow. Thanks.

Commit:
PR feedback - output datetime in iso format instead of java native format

johnnonweiler commented on data_pipeline_api/java_api/src/main/java/uk/ramp/config/ConfigFactory.java:
```diff
@@ -0,0 +1,30 @@
+            .hash(fileDirectoryNormaliser.normalisePath(LOCATION));
+    var runId = config.runId().orElse(freshHash);
+    return config
+        .withFailOnHashMisMatch(failOnHashMismatch)
```
Why do you set failOnHashMismatch here, when it is already set?

vinopm replied:
The `failOnHashMismatch` is optional and can be left out in the config file. So a few lines above I am loading the optional value for `failOnHashMismatch` and defaulting it to false and then inserting it back into the config, to ensure it is not empty after the config is loaded

vinopm replied:
I've removed the setting of failOnHashMismatch here now. Please see your comment about 

'With the immutable/json/yaml libraries you're using, is it possible to have a boolean with a default value of true here, rather than using an Optional <Boolean>?' 

for more info.

johnnonweiler commented on data_pipeline_api/java_api/src/main/java/uk/ramp/file/FileReader.java:
```diff
@@ -0,0 +1,16 @@
+    try (java.io.FileReader fileReader = new java.io.FileReader(new File(fileName))) {
+      return IOUtils.toString(fileReader);
+    } catch (IOException ex) {
+      throw new UncheckedIOException(ex);
```
Hmmm... it's interesting that you don't create a new IOException in this case.  (There seems to be something I'm missing here!)

vinopm replied:
See my other comment! The usage here is correct, the other one was a mistake

johnnonweiler commented on data_pipeline_api/java_api/src/main/java/uk/ramp/metadata/HashVerificationMetadataSelector.java:
```diff
@@ -0,0 +1,41 @@
+
+    var verifiedHash = readMetaDataItem.verifiedHash().get();
+
+    String calculatedHash = hasher.hash(readMetaDataItem.filename().get());
```
Was there a particular reason why you didn't use `.orElseThrow()` here, instead of checking `isEmpty()` then using `get()`?

vinopm replied:
No particular reason, updated it now to use `.orElseThrow()` - saves a few lines of code now

johnnonweiler commented on data_pipeline_api/java_api/src/main/java/uk/ramp/config/Config.java:
```diff
@@ -0,0 +1,71 @@
+  }
+
+  @JsonProperty("fail_on_hash_mismatch")
+  Optional<Boolean> failOnHashMisMatch();
```
With the immutable/json/yaml libraries you're using, is it possible to have a boolean with a default value of *true* here, rather than using an Optional \<Boolean\>?

vinopm replied:
Yes, I will change this to:

```
  @JsonProperty("fail_on_hash_mismatch")
  Optional<Boolean> internalFailOnHashMisMatch();

  @Derived
  default boolean failOnHashMisMatch() {
    return internalFailOnHashMisMatch().orElse(true);
  }
```

johnnonweiler commented on data_pipeline_api/java_api/src/main/java/uk/ramp/metadata/MatchingMetadataSelector.java:
```diff
@@ -0,0 +1,21 @@
+  public MetadataItem find(MetadataItem key) {
+    return metadataItems.stream()
+        .filter(i -> i.isSuperSetOf(key))
+        .max(comparing(i -> i.version().orElse("")))
```
Are you confident that this comparison is going to give the correct ordering?  It looks like it is using a string comparison where it should be using a numeric comparison.  I suggest adding version "11.0.0" to your test cases.

I think there has been some discussion about what goes into version strings on zulip, but I don't know what the conclusion was.

vinopm replied:
Good point. I initially thought the Python implementation was doing this naively too, but just double checked and it's using a library to do the comparison. I've updated this now to make sure we are doing a proper comparison. Please review again. Thanks

johnnonweiler commented on data_pipeline_api/java_api/src/test/java/uk/ramp/metadata/MatchingMetadataSelectorTest.java:
```diff
@@ -0,0 +1,57 @@
+          .build();
+
+  @Before
+  public void setUp() {
```
I think this setUp() could be deleted.

vinopm replied:
Updated

johnnonweiler commented on data_pipeline_api/java_api/src/main/java/uk/ramp/metadata/MetadataItem.java:
```diff
@@ -0,0 +1,108 @@
+  Optional<String> source();
+
+  default boolean isSuperSetOf(MetadataItem key) {
+    return key.filename().map(k -> k.equals(filename().orElse(""))).orElse(true)
```
You could a method here, not sure what to call it, to do:
```
boolean keyIsEitherNotPresentOrEqual(Optional<String> key, Optional<String> other) {
    return key.map(k -> k.equals(superset.orElse(""))).orElse(true);
}
```

vinopm replied:
Updated this, hopefully is more clear now. I generally try to avoid passing in Optionals through parameters as it is considered bad practice by some. So I have passed in empty string instead, doing something quite similar to your suggestion. Let me know what you think.

johnnonweiler commented on data_pipeline_api/java_api/src/main/java/uk/ramp/metadata/MetadataItem.java:
```diff
@@ -0,0 +1,108 @@
+            .filter(
+                readOverrideItem ->
+                    this.isSuperSetOf(
+                        readOverrideItem.where().orElse(ImmutableMetadataItem.copyOf(this))))
```
Is `this.isSuperSetOf(
                        readOverrideItem.where().orElse(ImmutableMetadataItem.copyOf(this)))`
equivalent to
`readOverrideItem.where().map(k -> this.isSuperSetOf(k)).orElse(true)` ?

vinopm replied:
Yes, it is. Much nicer, updated. Thanks.

johnnonweiler reviewed (commented):
This looks like great stuff.  You have clearly been very busy!

I have only got half way through reading this, but have to pause for a bit now, so I'll share what I've got so far.  I have been trying to read it carefully - most of my comments or questions are very minor things, and could be ignored.

johnnonweiler commented on data_pipeline_api/java_api/src/main/java/uk/ramp/metadata/MetadataSelectorFactory.java:
```diff
@@ -0,0 +1,20 @@
+        new ArrayList<>(new MetaDataReader(yamlReader, fileDirectoryNormaliser).read());
+    MetadataSelector baseMetadataSelector = new MatchingMetadataSelector(metadataItems);
+    return new HashVerificationMetadataSelector(baseMetadataSelector, hasher, shouldVerifyHash);
+  }
```
Should AccessLoggedMetadataSelector be used somewhere ?

vinopm replied:
Moved this up to the FileApi level now so that it is more visible

johnnonweiler commented on data_pipeline_api/java_api/src/main/java/uk/ramp/overrides/OverridesApplier.java:
```diff
@@ -0,0 +1,49 @@
+  }
+
+  public MetadataItem applyWriteOverrides(MetadataItem query) {
+    List<OverrideItem> readOverrides = new ArrayList<>(config.writeQueryOverrides());
```
I suggest `readOverrides` is renamed - there are two words in English which are spelt 'read', but pronounced differently, and this potentially causes confusion here.
(Could `config.writeQueryOverrides()` be used directly below, instead of `readOverrides`?)

vinopm replied:
Agreed, updated.

johnnonweiler commented on data_pipeline_api/java_api/src/main/java/uk/ramp/overrides/OverridesApplier.java:
```diff
@@ -0,0 +1,49 @@
+    return query
+        .applyOverrides(readOverrides)
+        .applyOverrides(List.of(runIdOverride))
+        .applyOverrides(List.of(filenameOverride));
```
This appears to be sometimes overriding the filename with its original value.  Is this the intention?  Is it deliberately overwriting filename values from the config file?

vinopm replied:
We are overriding the filename with a generated file name (based on the data product, run id, and extension provided in the query) if the filename is not provided in the query. If the file name is provided, we just override the filename with the same filename (which has no effect). The config file is not used here to determine the filename override. 

Maybe I am missing the point you were making, could you clarify please if that's the case. Thanks!

johnnonweiler commented on data_pipeline_api/java_api/src/main/java/uk/ramp/overrides/OverridesApplier.java:
```diff
@@ -0,0 +1,49 @@
+  }
+
+  public MetadataItem applyReadOverrides(MetadataItem query) {
+    return query.applyOverrides(new ArrayList<>(config.readQueryOverrides()));
```
Why is `new ArrayList<>` needed here? (I'm just curious.)

vinopm replied:
`config.readQueryOverrides()` has the type `ImmutableOverrideItem`. The `applyOverrides()` method takes in type `List<OverrideItem>` instead of `List<ImmutableOverrideItem>`. `OverrideItem` is an interface of `ImmutableOverrideItem`, so I've done this so that the types match up.

johnnonweiler commented on data_pipeline_api/java_api/src/main/java/uk/ramp/reader/Reader.java:
```diff
@@ -0,0 +1,5 @@
+package uk.ramp.reader;
+
+public interface Reader<T> {
```
What is the purpose of the `Reader<T>` and `Writer<T>` interfaces?
(Is it a kind of documentation thing?)

vinopm replied:
Removed these two interfaces, they weren't necessary. Thanks.

johnnonweiler commented on data_pipeline_api/java_api/src/test/java/uk/ramp/file/FileDirectoryNormaliserTest.java:
```diff
@@ -0,0 +1,28 @@
+  }
+
+  @Test
+  public void normaliseRelativePath() {
```
Are the names of these two tests the wrong way round?  This test is normalising an absolute path, and the following test is normalising a relative path.

vinopm replied:
You're correct. I've updated now. Thanks.

Commit:
PR feedback update - Ensure that we only add write access log after the file handle for writing has been closed. Also added a safety net for making sure access log is written out even if .close() is not explicitly called by the client.

johnnonweiler commented on data_pipeline_api/java_api/src/test/resources/config.yaml:
```diff
@@ -0,0 +1,23 @@
+data_directory: data
+access_log: access-{run_id}.yaml
```
This access_log line looks odd, but I guess it doesn't make any difference to anything.

vinopm replied:
@johnnonweiler Copied this from the original python implementation. Seems that in the python implementation they replace `{run_id}` with the actual run id. @rtblackwell @mrow84 is that correct?

mrow84 replied:
Yes, there is a description of the mechanism in the [file API spec](https://teams.microsoft.com/l/file/8B530C1E-25FE-4896-BD0C-8FC8089C5AC7?tenantId=6e725c29-763a-4f50-81f2-2e254f0133c8&fileType=docx&objectUrl=https%3A%2F%2Fgla.sharepoint.com%2Fsites%2FScottishCOVID-19ResearchConsortium%2FShared%20Documents%2FGeneral%2FCollaboration%2Fdata%20pipeline%20API%2F1.%20File%20API.docx&baseUrl=https%3A%2F%2Fgla.sharepoint.com%2Fsites%2FScottishCOVID-19ResearchConsortium&serviceName=teams&threadId=19:283c5ea4551344249cc76dd0ecaa4f74@thread.tacv2&groupId=669b45e2-a9a6-4060-ba8a-9ebd3713e367) (link may or may not work, file is in teams at: General > Collaboration > data pipeline API).

johnnonweiler replied:
Might there be any other instances of {run_id} in the yaml file that should be replaced, or is this just for the access_log?

vinopm replied:
I believe it's an edge case specific to the access_log. The current python implementation only carries out this replacement with the access_log. @mrow84 can you confirm if this is true? Thanks

mrow84 replied:
The underlying feature is to associate output files with the run that produced them. There are two categories of output file: those produced by the model, handled by the file API write mechanism, and the access log. Both categories are identified using the run id - the former during the `open_for_write` process, and the latter as this transformation of the input filename. Ideally you _always_ want the run id in the access log filename, but we allow for the case where that isn't so (which can be useful).

I don't think the transformation is required elsewhere.

vinopm replied:
Thanks Chris.

Commit:
PR feedback - simplify HasherFactory

johnnonweiler reviewed (commented):
It looks like you've updated everything while I've been reviewing it, so my comments are probably all out of date.

I'm concerned about the comparison of version numbers, but otherwise this all looks pretty good.

I believe you have made changes only within the java_api folder, but there are files in the top level of data_pipeline_api which will probably want updating (e.g. README.md).

I wonder if we should be thinking about targetting this to the Java version on the DiRAC HPC system.  (I think it is currently using some newer features of Java which might not be available there.)

vinopm commented:
> It looks like you've updated everything while I've been reviewing it, so my comments are probably all out of date.
> 
> I'm concerned about the comparison of version numbers, but otherwise this all looks pretty good.
> 
> I believe you have made changes only within the java_api folder, but there are files in the top level of data_pipeline_api which will probably want updating (e.g. README.md).
> 
> I wonder if we should be thinking about targetting this to the Java version on the DiRAC HPC system. (I think it is currently using some newer features of Java which might not be available there.)

@johnnonweiler Haven't changed much, just updated mainly to reflect feedback from Chris. Please continue reviewing, your feedback so far has been great!

Rikiya has been able to run the Contact Tracing Model with Java 11 on DiRAC, so i suppose we could do the same with the Covid Sim Model?

Commit:
PR feedback update - More tests as requested by John

Commit:
PR feedback update - File normaliser make sure that the directories created are based on the final parent path rather than what we have stored in the normaliser as a parent path.

Commit:
PR feedback update - Pass in normalised path into ConfigReader instead of the normaliser, and declare the static location in the factory only.

Commit:
Clean up

Commit:
PR feedback update - Simplify use of isEmpty/get/orElseThrow

Commit:
PR feedback update - Set default value for failOnHashMismatch within Config interface

Commit:
PR feedback update - Use library for version comparison + tests

Commit:
PR feedback update - Cleanup

Commit:
PR feedback update - Clean up meta data is super set of

Commit:
PR feedback update - meta data item tests

Commit:
Fix conflicting jackson mapping

Commit:
Clean up

Commit:
PR feedback update - use config.writeQueryOverrides directly instead of readOverrides

Commit:
Fix test names

vinopm commented on data_pipeline_api/java_api/src/test/java/uk/ramp/hash/Sha1HasherTest.java:
```diff
@@ -0,0 +1,18 @@
+
+public class Sha1HasherTest {
+
+  @Test
+  public void testSha1Hash() {
+    var input = "test";
+
+    // Generated via http://www.sha1-online.com/
+    var expectedOutput = "a94a8fe5ccb19ba61c4c0873d391e987982fbbd3";
+
+    assertThat(new Sha1Hasher().hash(input)).isEqualTo(expectedOutput);
+  }
```
@johnnonweiler sha1 test is here now

Commit:
PR feedback update - remove unnecessary generic interface Reader.java and Writer.java

Commit:
PR feedback update - simplify stream logic

vinopm reviewed (commented):
Updated PR. Please could you review again @johnnonweiler @rtblackwell @mrow84. Thanks

johnnonweiler commented on data_pipeline_api/java_api/src/main/java/uk/ramp/file/CleanableFileChannel.java:
```diff
@@ -0,0 +1,150 @@
+
+  @Override
+  public long read(ByteBuffer[] dsts) throws IOException {
+    return 0;
```
This looks a bit different to all the other overridden methods.  Why is this not doing a read?

vinopm replied:
Thanks for spotting this mistake. I've updated it.

johnnonweiler reviewed (approved):
Looks good to me.  Thanks for the thorough job addressing all of my previous comments.

johnnonweiler commented:
I'm happy with Java 11, if Rikiya has got it working on DiRAC.  (And would be interested to know how he did that.)

Commit:
PR feedback update - fix bug + return CleanableFileChannel instead of FileChannel to maintain encapsulation

vinopm commented:
@johnnonweiler  Fixed the bug you pointed out + updated two other methods to return the instance of `CleanableFileChannel` instead of the underlying `FileChannel` in order to maintain encapsulation. Please review the small change. Thanks



> I'm happy with Java 11, if Rikiya has got it working on DiRAC. (And would be interested to know how he did that.)

Yes, he has managed to run with java 11 on DiRAC. I believe he downloaded OpenJDK 11 to his DiRAC home file store, and then pointed the `$JAVA_HOME` env variable to the downloaded JDK.

johnnonweiler reviewed (approved):
Looks good to me. 

Why did you add `if (fileChannel == null) { return null; }` ?  It appears from the FileChannel documentation that I looked at that `fileChannel` should never be null.

vinopm commented:
> Looks good to me.
> 
> Why did you add `if (fileChannel == null) { return null; }` ? It appears from the FileChannel documentation that I looked at that `fileChannel` should never be null.

Thanks @johnnonweiler. The documentation doesn't mention it but when I had a look at the implementation, it's returning null in some cases: http://hg.openjdk.java.net/jdk/jdk11/file/tip/src/java.base/share/classes/sun/nio/ch/FileChannelImpl.java#l424 Have a look at lines 359, 424, 433.

johnnonweiler commented:
So, the documentation could do with some improvement, and it's clearly a good idea to look at their implementation!

Commit:
1. Change API so that we pass in the actual config file path into constructor of API instead of the folder path
2. some refactoring
3. Added JsonIgnore annotations to make sure derived fields are not written out to the access_log file
4. Previously files were not using the data_direcory parent path, they were using the config folder. This is fixed now.

vinopm commented:
> So, the documentation could do with some improvement, and it's clearly a good idea to look at their implementation!

Yes exactly!

vinopm commented:
@johnnonweiler I've realised I needed to make a few changes after giving the data API document a thorough read. Please could you review my latest changes:

1. Change API so that we pass in the actual config file path into constructor of API instead of the folder path
2. some refactoring
3. Added JsonIgnore annotations to make sure derived fields are not written out to the access_log file
4. Previously the access_log was being written out to the data_directory folder instead of using the access_log path properly. When it was specified as a relative path, it was writing out into the data_directory folder. This is fixed now so that we reference the parent path to be the same folder that the config file is in.

Hopefully this is the last set of changes for this PR, if my changes in the last 3 commits look ok

Thanks again

Commit:
To fix CI complaint

Commit:
Add documentation to explain why we are checking for and returning null

johnnonweiler reviewed (approved):
Looks good to me.  (I think there _might_ be cases where you're returning a String, and it would make more sense to return a Path, but it is good as it is.)

johnnonweiler commented:
Incidentally, is there a plan to update files in the top level of data_pipeline_api, such as README.md, in a future PR?

vinopm commented:
@johnnonweiler agreed, I'll address standardising the return types to `Path` in the next PR. I'll update the README.md in a future PR as well.

@mrow84 @rtblackwell could we have this merged? 

Thanks

mrow84 commented:
> Incidentally, is there a plan to update files in the top level of data_pipeline_api, such as README.md, in a future PR?

At the moment there is only [this ticket](https://github.com/ScottishCovidResponse/SCRCIssueTracking/issues/415) about adding example usage and [this ticket](https://github.com/ScottishCovidResponse/SCRCIssueTracking/issues/548) documenting the usage of the tagging mechanism (which is currently python specific, I think). We should certainly add more into the readme.

mrow84 commented:
I am slightly uncomfortable with the directory structure - I kind of feel like it might be better separated at the top level, but I am happy to do that later if you don't think it is going to conflict. In particular, I was wondering what will happen when we [package our code](https://github.com/ScottishCovidResponse/data_pipeline_api/blob/master/setup.py#L41) in python - @aflag will have a more informed opinion. You could try building the python package locally on your branch and checking what is included.

There are some new features that have been added since you started working on this, but I am in favour of getting this merged and adding those separately.

Commit:
move java api one level up

vinopm commented:
@mrow84 Moved it one level up now

vinopm commented:
@mrow84 Thanks for the approval. I will add the other changes such as functionality for `set_metadata` in a follow up PR. Could you go ahead and merge if you're comfortable. I don't have permissions to merge. Thanks
